idx/metric_definition.go

```diff
@@ -0,0 +1,546 @@
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	// get []int of the lengths of all of the mn.Nodes
+	lns, ok := IdxIntern.Len(mn.nodes)
+	if !ok {
+		internError.Inc()
+		log.Error("idx: Failed to retrieve length of strings from interning library for MetricName")
+		return ""
+	}
+
+	// should be faster than calling IdxIntern.SetString in a tight loop
+	var tmpSz string
+	szHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpSz))
+	first, _ := IdxIntern.GetStringFromPtr(mn.nodes[0])
+	bld.WriteString(first)
+	for idx, nodePtr := range mn.nodes[1:] {
+		szHeader.Data = nodePtr
+		szHeader.Len = lns[idx+1]
+		bld.WriteString(".")
+		bld.WriteString(tmpSz)
+	}
+
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) {
+	nodes := strings.Split(name, ".")
+	mn.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node), false)
+		if err != nil {
+			log.Error("idx: Failed to acquire interned string for node name: ", err)
+			internError.Inc()
+		}
+		mn.nodes[i] = nodePtr
+	}
+}
+
+// ExtensionType is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) ExtensionType() int8 {
+	return 95
+}
+
+// Len is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) Len() int {
+	return len(mn.String())
+}
+
+// MarshalBinaryTo is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) MarshalBinaryTo(b []byte) error {
+	copy(b, []byte(mn.String()))
+	return nil
+}
+
+// UnmarshalBinary is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) UnmarshalBinary(b []byte) error {
+	mn.setMetricName(string(b))
+	return nil
+}
+
+//msgp:shim TagKeyValue as:string using:TagKeyValue.createTagKeyValue/parseTagKeyValue
+
+// TagKeyValue stores a Key/Value pair. The strings
+// are interned in an object store before they are assigned.
+type TagKeyValue struct {
+	Key   uintptr
+	Value uintptr
+}
+
+// String returns a Key/Value pair in the form of
+// 'key=value'
+func (t *TagKeyValue) String() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
```

Comment:
I think *anywhere* we fail to reconstruct a string from the interning layer, we should *not* return an empty string and *not* return whatever the interning library gave us.
instead return "intern-error" or something

Reply:
actually, since in so many places we call an intern layer method, check error, and if so increment the error counter and log an error..
perhaps we should have a small wrapper that handles all that, and returns "intern-error" string in case of an error? at least for those calls where we can recover from errors of course.
this would keep the idx code cleaner

Reply:
If we fail to retrieve an already interned string from the interning layer then something has gone seriously wrong. Do we instead just return an error and have the caller check for errors?

If we are missing a tag from a MetricDefinition is that something we can actually recover from?

Reply:
> Do we instead just return an error and have the caller check for errors?

how could the caller handle an error that happened inside of the interning layer?

> If we are missing a tag from a MetricDefinition is that something we can actually recover from?

 sure, it just means when you issue a render request, your series will have missing values in their tags, and maybe that will affect aggregations etc, but we have to make the choice to handle it like that or fail the entire request. I don't think we should do the latter.

obviously we do have to make sure to never persist anything that we could not properly construct

Reply:
As far as how the caller would handle an error that happened inside of the interning layer... I was thinking we just fail whatever they were doing, or continue to pass the error up to the next higher level. It depends on who the caller is. 

The interning layer needs more rigorous testing, but if it is not returning data for a given `uintptr` it will continue to not return data. If at some point it does begin returning data we can also assume that it is not the correct data, but more than likely something else was added to the object store at that address and it is now returning that value. This is even worse, because we will be receiving incorrect data and thinking that everything is fine. I think failing and invalidating as early as possible is the best way to handle this.

Reply:
So to handle the case you're describing we have no choice but to crash the process.

Reply:
Unless we can come up with a way to recover from it, yes.

Reply:
I think it would be better to handle this like in the following `createTagKeyValue()` function and then alert on that metric which counts intern errors.